src/ui/controllers/voicesearchcontroller.js

```diff
@@ -0,0 +1,135 @@
+     * The screen reader text for the "Stop Voice Search" button
+     * @type {string}
+     */
+    this._stopVoiceSearchText = config.stopText || TranslationFlagger.flag({
+      phrase: 'Stop Voice Search',
+      context: 'A button to stop listening for a voice search'
+    });
+
+    this._searchBarContainer = searchBarContainer;
+    this._voiceSearchElement = DOM.query(searchBarContainer, '.yxt-SearchBar-voiceSearch');
+    this._screenReaderTextElement = DOM.query(searchBarContainer, '.yxt-SearchBar-voiceSearchText');
+    this._updateScreenReaderText();
+
+    /** @type {SVGAnimateElement} */
+    this._micFadeInAnimation = DOM.query(searchBarContainer, '.js-yxt-SearchBar-micFadeIn');
+    /** @type {SVGAnimateElement} */
+    this._micFadeOutAnimation = DOM.query(searchBarContainer, '.js-yxt-SearchBar-micFadeOut');
+    /** @type {SVGAnimateElement[]} */
+    this._dotsFadeInAnimations = DOM.queryAll(searchBarContainer, '.js-yxt-SearchBar-dotsFadeIn');
+    /** @type {SVGAnimateElement[]} */
+    this._dotsFadeOutAnimations = DOM.queryAll(searchBarContainer, '.js-yxt-SearchBar-dotsFadeOut');
+
+    this._micIconWrapper = DOM.query(this._searchBarContainer, '.yxt-SearchBar-micIconWrapper');
+    this._listeningIconWrapper = DOM.query(this._searchBarContainer, '.yxt-SearchBar-listeningIconWrapper');
+
+    this._customMicIconUrl = config.customMicIconUrl;
+    this._customListeningIconUrl = config.customListeningIconUrl;
+
+    if (this._customMicIconUrl) {
+      this._applyCustomMicIconStyling();
+    }
+
+    if (this._customListeningIconUrl) {
+      this._applyCustomListeningIconStyling();
+    }
+  }
+
+  handleIconClick () {
+    if (this._state === State.NOT_LISTENING) {
+      this._state = State.LISTENING;
+    } else if (this._state === State.LISTENING) {
+      this._state = State.NOT_LISTENING;
+    }
+    this._updateUI();
+  }
+
+  _updateUI () {
+    if (this._customMicIconUrl) {
+      this._applyCustomMicIconStyling();
+    } else {
+      this._executeDefaultMicIconAnimations();
+    }
+    if (this._customListeningIconUrl) {
+      this._applyCustomListeningIconStyling();
+    } else {
+      this._executeDefaultListeningIconAnimations();
+    }
+    this._updateScreenReaderText();
+    this._voiceSearchElement.blur();
+  }
+
+  _executeDefaultMicIconAnimations () {
+    if (this._state === State.NOT_LISTENING) {
+      this._micFadeInAnimation.beginElement();
+    } else if (this._state === State.LISTENING) {
+      this._micFadeOutAnimation.beginElement();
+    }
+  }
+
+  _executeDefaultListeningIconAnimations () {
+    if (this._state === State.NOT_LISTENING) {
+      this._dotsFadeOutAnimations.forEach(animation => {
+        animation.beginElement();
+      });
+    } else if (this._state === State.LISTENING) {
+      this._dotsFadeInAnimations.forEach(animation => {
+        animation.beginElement();
+      });
+    }
+  }
+
+  _applyCustomMicIconStyling () {
+    if (this._state === State.NOT_LISTENING) {
```

Comment:
would it be weird if we just toggled one css class on the `yxt-SearchBar-voiceIconWrapper` container, instead of toggling these classes on the individual divs?

Reply:
not sure which way would be cleaner overall, on one hand I think just toggling one css class is nice, but the css also has to "know more" in that case

Reply:
I tried the way you suggested and you're right, the javascript is a little simpler but the styling is a little more complicated. I don't really have a preference as it can really go either way